### PR TITLE
[FrameworkBundle] Define APP_ENV/APP_DEBUG from argv via Application::bootstrapEnv()

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28984, #29126
| License       | MIT
| Doc PR        | todo

Replaces #29126.

### CODE REVIEW

File: UPGRADE-4.2.md [hunk 1]
```diff
@@ -175,10 +175,10 @@ FrameworkBundle
    ```
  * The `ContainerAwareCommand` class has been deprecated, use `Symfony\Component\Console\Command\Command`
    with dependency injection instead.
- * The `--env` console option and its "-e" shortcut have been deprecated,
-   set the "APP_ENV" environment variable instead.
- * The `--no-debug` console option has been deprecated,
-   set the "APP_DEBUG" environment variable to "0" instead.
+ * The `--env` console option and its "-e" shortcut havee been deprecated, set the "APP_ENV" environment variable
```
Comment (ro0NL) [inline UPGRADE-4.2.md]: `have` :)

File: UPGRADE-5.0.md [hunk 2]
```diff
@@ -165,9 +165,9 @@ FrameworkBundle
  * The `ContainerAwareCommand` class has been removed, use `Symfony\Component\Console\Command\Command`
    with dependency injection instead.
  * The `--env` console option and its "-e" shortcut have been removed,
-   set the "APP_ENV" environment variable instead.
- * The `--no-debug` console option has been removed, 
-   set the "APP_DEBUG" environment variable to "0" instead.
+   set the "APP_ENV" environment variable instead or use `Symfony\Component\Console\Application::bootstrapEnv()` instead.
```
Comment (nicolas-grekas) [inline UPGRADE-5.0.md]: the first "instead" in the line can be removed

File: src/Symfony/Component/Console/Application.php [hunk 3]
```diff
@@ -1091,6 +1091,35 @@ public function setDefaultCommand($commandName, $isSingleCommand = false)
         return $this;
     }
 
+    /**
+     * Defines the "APP_ENV" and "APP_DEBUG" environment variables from --env and --no-debug argv options.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Application.php]: Defines the "APP_ENV" and "APP_DEBUG" environment variables by consuming --env and --no-debug from the command line arguments.

File: src/Symfony/Component/Console/Application.php [hunk 4]
```diff
@@ -1091,6 +1091,35 @@ public function setDefaultCommand($commandName, $isSingleCommand = false)
...
+     */
+    public static function bootstrapEnv(array &$argv)
+    {
+        for ($i = 0; $i < \count($argv) && '--' !== $v = $argv[$i]; ++$i) {
+            if ('--no-debug' === $v) {
+                putenv('APP_DEBUG='.(int) $_SERVER['APP_DEBUG'] = $_ENV['APP_DEBUG'] = false);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Application.php]: putenv('APP_DEBUG='.$_SERVER['APP_DEBUG'] = $_ENV['APP_DEBUG'] = '0');

**Review by nicolas-grekas — COMMENTED** [review]
cool thanks :)

File: UPGRADE-4.2.md (hunk 1 above)
Comment (chalasr) [reply UPGRADE-4.2.md]: fixed thanks :)

**Comment by chalasr** [comment]
@nicolas-grekas updated, thanks

**Review by nicolas-grekas — APPROVED** [review]
thanks!

File: src/Symfony/Component/Console/Application.php [hunk 5]
```diff
@@ -1091,6 +1091,35 @@ public function setDefaultCommand($commandName, $isSingleCommand = false)
         return $this;
     }
 
+    /**
+     * Defines the "APP_ENV" and "APP_DEBUG" environment variables by consuming --env and --no-debug from the command line arguments.
+     */
+    public static function bootstrapEnv(array &$argv)
```
Comment (ro0NL) [inline src/Symfony/Component/Console/Application.php]: i think adding it to `Symfony\Bundle\FrameworkBundle\Console\Application` makes more sense, the logic seems tied to the framework stack; not the component.

`FWBundle\Application` is already the used class in bin/console.. so should be fine no?

Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Application.php]: Adding it to the component allows standardizing the practice. And the code has not relationships with the bundle. My vote is to keep it in the component.

Comment (ro0NL) [reply src/Symfony/Component/Console/Application.php]: Right :+1: seems fair.

Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Application.php]: but ok to put in FWB if more ppl think it's a better fit :)

**Review by nicolas-grekas — APPROVED** [review]
OK, I agree, too opinionated for a component :)

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 6]
```diff
@@ -15,10 +15,10 @@ CHANGELOG
  * Removed the `framework.messenger.encoder` and `framework.messenger.decoder` options. Use the `framework.messenger.serializer.id` option to replace the Messenger serializer. 
  * Deprecated the `ContainerAwareCommand` class in favor of `Symfony\Component\Console\Command\Command`
  * Made `debug:container` and `debug:autowiring` ignore backslashes in service ids
- * Deprecated the `--env` console option and its "-e" shortcut, set
-   the "APP_ENV" environment variable instead.
- * Deprecated the `--no-debug` console option, set the "APP_DEBUG" 
-   environment variable to "0" instead.
+ * Deprecated the `--env` console option and its "-e" shortcut, set the "APP_ENV" environment variable
+   or use `Symfony\Component\Console\Application::bootstrapEnv()` instead.
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: FWBundle namespace ;)

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 7]
```diff
@@ -15,10 +15,10 @@ CHANGELOG
...
+ * Deprecated the `--no-debug` console option, set the "APP_DEBUG" environment variable to "0"
+   or use `Symfony\Component\Console\Application::bootstrapEnv()` instead.
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: same

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md (hunk 6 above)
Comment (chalasr) [reply src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: thanks, fixed

**Comment by nicolas-grekas** [comment]
I'm merging quickly so that the recipes work with 4.2 (they're broken for now.)
It will allow us to play with the new experience more easilly to ensure 4.2 will be smooth.
If you have more comments here, please do post them and I'll ensure they're resolve asap.

**Comment by nicolas-grekas** [comment]
Thank you @chalasr.

**Comment by fmata** [comment]
I fixed a minor doc issue + removed `\count()` calls in #29164.

**Pull request merged**